alloydb/notebooks/embeddings_batch_processing.ipynb

```diff
@@ -0,0 +1,1160 @@
+    "      task_type: The task type for the embedding model. Defaults to\n",
+    "        \"SEMANTIC_SIMILARITY\".\n",
+    "        Supported task types: https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/task-types\n",
+    "      retries: The maximum number of times to retry embedding generation in case\n",
+    "        of errors. Defaults to 100.\n",
+    "      delay: The delay in seconds between retries. Defaults to 30.\n",
+    "\n",
+    "    Returns:\n",
+    "      A list of records containing ids and embeddings.\n",
+    "      Example:\n",
+    "        [\n",
+    "          {\n",
+    "            'id': 'id1',\n",
+    "            'col1_embedding': [1.0, 1.1, ...],\n",
+    "            'col2_embedding': [2.0, 2.1, ...],\n",
+    "            ...\n",
+    "          },\n",
+    "          ...\n",
+    "        ]\n",
+    "      where col1 and col2 are columns containing data to be embedded.\n",
+    "    Raises:\n",
+    "      Exception: Raises the encountered exception if all retries fail.\n",
+    "    \"\"\"\n",
+    "    logger = logging.getLogger(\"embed_objects\")\n",
+    "    global total_char_count\n",
+    "\n",
+    "    # Place all of the embeddings into a single list\n",
+    "    inputs = []\n",
+    "    for row in batch_data:\n",
+    "        for col in cols_to_embed:\n",
+    "            if col in row and row[col]:\n",
+    "                inputs.append(TextEmbeddingInput(row[col], task_type))\n",
+    "\n",
+    "    # Retry loop\n",
+    "    for attempt in range(retries):\n",
+    "        try:\n",
+    "            # Get embeddings for the text data\n",
+    "            embeddings = await model.get_embeddings_async(inputs)\n",
+    "\n",
+    "            # Increase total char count\n",
+    "            total_char_count += sum([len(input.text) for input in inputs])\n",
+    "\n",
+    "            # group the results together by id\n",
+    "            embedding_iter = iter(embeddings)\n",
+    "            results = []\n",
+    "            for row in batch_data:\n",
+    "                r = {\"id\": row[\"id\"]}\n",
+    "                for col in cols_to_embed:\n",
+    "                    if col in row and row[col]:\n",
+    "                        r[f\"{col}_embedding\"] = str(next(embedding_iter).values)\n",
+    "                    else:\n",
+    "                        r[f\"{col}_embedding\"] = None\n",
+    "                results.append(r)\n",
+    "            return results\n",
+    "\n",
+    "        except Exception as e:\n",
+    "            if attempt < retries - 1:  # Retry only if attempts are left\n",
+    "                logger.warning(f\"Error: {e}. Retrying in {delay} seconds...\")\n",
+    "                await asyncio.sleep(delay)  # Wait before retrying\n",
+    "            else:\n",
+    "                logger.error(f\"Failed to get embeddings after {retries} attempts.\")\n",
+    "                raise  # Raise the error if all retries fail\n",
```

Comment:
When we log this, can you log _which_ embedding failed?

I think we may also want to skip raising the error so the rest of the pipeline continues.